pkg/api/errors.go

```diff
@@ -1,44 +1,75 @@
 package api
 
-import "errors"
-
-var (
-	Error400     = errors.New("Bad Request")
-	Error401     = errors.New("Not Authorized")
-	Error403     = errors.New("Forbidden")
-	Error404     = errors.New("Resource Not Found")
-	Error409     = errors.New("Conflict")
-	Error412     = errors.New("Precondition Failed")
-	Error429     = errors.New("Too Many Requests")
-	Error500     = errors.New("API Internal Error")
-	ErrorUnknown = errors.New("Unknown API Error")
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 )
 
-func getStatusError(code int) error {
-	switch code {
-	case 200:
-		return nil
-	case 202:
+type APIError struct {
+	Error struct {
+		Status    int            `json:"status"`
+		Message   string         `json:"message"`
+		Errors    []ErrorMessage `json:"errors"`
+		Reference string         `json:"reference"`
+		Source    string         `json:"source"`
+	} `json:"error"`
+}
+
+type ErrorMessage struct {
+	Message string `json:"message"`
+}
+
+type BigAnimalError struct {
+	APIError APIError
+	Err      error
+}
+
+func (baerr *BigAnimalError) Error() string {
+	return fmt.Sprintf("status: %d - %v", baerr.APIError.Error.Status, baerr.APIError.Error.Message)
+}
+
+func (baerr *BigAnimalError) getDetails() string {
+	var details string
+	for _, err := range baerr.APIError.Error.Errors {
+		details += fmt.Sprintln(err.Message)
+	}
+	return details
+}
+
+func FromErr(err error) diag.Diagnostics {
```

Comment:
api.FromErr tries to assert the BigAnimalError type to the error. 
If this succeeds, it returns a crafted diag.Diagnostics 
If not, it calls the `diag.FromErr` function. 

I've changed the `diag.FromErr` to `api.FromErr` function only after the Client function calls, which call the `doRequest` function. 
Technically, any `diag.FromErr` function can be changed with `api.FromErr` for the sake of expediency. @nstott what do you think about that? I don't have a strong opinion on that.

Reply:
tbh, I think we should avoid embedding the diag's into the api package.

the library code should try to return standard golang error types.  `BigAnimalError` is a good way to do that

we can compare error types with error `Is()` and `As()` https://gosamples.dev/check-error-type/ https://pkg.go.dev/errors#As

errors.Wrap might also be useful if we are chaining errors

Reply:
100% agree with keeping the api package clean, without the sdk/diag import. 
I've moved the FromErr function to `pkg/provider/utils.go` and named it `BigAnimalFromErr`.(any better naming suggestions are welcome, I couldn't come up with something better 😅)
I've written the comparison with `As()`.